### PR TITLE
[LinearSystem] Assemble non-mapped and mapped matrices in parallel

### PR DESCRIPTION
**Pull request by alxbilger**
Based on #4158 

Non-mapped and mapped matrices can be assembled in parallel because the contributions are not accumulated in the same matrix data structure.

https://github.com/alxbilger/sofa/compare/constantsparsitypattern...alxbilger:sofa:independantmatrices

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

**Comment by alxbilger** [comment]
[ci-build][with-all-tests]

**Comment by alxbilger** [comment]
From the scene `examples/Component/LinearSystem/MatrixLinearSystem.scn`

The timeline of the sequential assembly:
assembleSequential


The timeline of the parallel assembly:
assembleParallel

Both pictures are not at the same scale

**Comment by alxbilger** [comment]
Since this PR is based on an another, it can be difficult to review it. Here is the diff with the other PR: https://github.com/alxbilger/sofa/compare/constantsparsitypattern...alxbilger:sofa:independantmatrices

File: Sofa/Component/LinearSystem/src/sofa/component/linearsystem/MatrixLinearSystem.h [hunk 1]
```diff
@@ -80,6 +81,7 @@ class MatrixLinearSystem : public TypedMatrixLinearSystem<TMatrix, TVector>
     Data< bool > d_applyProjectiveConstraints; ///< If true, projective constraints are applied on the global matrix
     Data< bool > d_applyMappedComponents; ///< If true, mapped components contribute to the global matrix
     Data< bool > d_checkIndices; ///< If true, indices are verified before being added in to the global matrix, favoring security over speed
+    Data< bool > d_parallelAssemblyIndependantMatrices; ///< If true, independant matrices (global matrix vs mapped matrices) are assembled in parallel
```
Comment (fredroy) [inline Sofa/Component/LinearSystem/src/sofa/component/linearsystem/MatrixLinearSystem.h]:
```suggestion
    Data< bool > d_parallelAssemblyIndependentMatrices; ///< If true, independent matrices (global matrix vs mapped matrices) are assembled in parallel
```

https://en.wiktionary.org/wiki/independant 🙊

File: Sofa/Component/LinearSystem/src/sofa/component/linearsystem/MatrixLinearSystem.inl [hunk 2]
```diff
@@ -59,6 +61,8 @@ MatrixLinearSystem<TMatrix, TVector>::MatrixLinearSystem()
     , d_applyProjectiveConstraints(initData(&d_applyProjectiveConstraints, true,  "applyProjectiveConstraints", "If true, projective constraints are applied on the global matrix"))
     , d_applyMappedComponents     (initData(&d_applyMappedComponents,      true,  "applyMappedComponents",      "If true, mapped components contribute to the global matrix"))
     , d_checkIndices              (initData(&d_checkIndices,               false, "checkIndices",               "If true, indices are verified before being added in to the global matrix, favoring security over speed"))
+    , d_parallelAssemblyIndependantMatrices
+        (initData(&d_parallelAssemblyIndependantMatrices, true, "parallelAssemblyIndependantMatrices", "If true, independant matrices (global matrix vs mapped matrices) are assembled in parallel"))
```
Comment (fredroy) [inline Sofa/Component/LinearSystem/src/sofa/component/linearsystem/MatrixLinearSystem.inl]:
```suggestion
    , d_parallelAssemblyIndependentMatrices
        (initData(&d_parallelAssemblyIndependentMatrices, true, "parallelAssemblyIndependentMatrices", "If true, independent matrices (global matrix vs mapped matrices) are assembled in parallel"))
```

File: Sofa/Component/LinearSystem/src/sofa/component/linearsystem/MatrixLinearSystem.inl [hunk 3]
```diff
@@ -116,26 +164,59 @@ void MatrixLinearSystem<TMatrix, TVector>::assembleSystem(const core::Mechanical
...
 
-        if (d_assembleStiffness.getValue())
-        {
-            contribute<Contribution::STIFFNESS>(mparams);
-        }
+        simulation::TaskScheduler* taskScheduler = simulation::MainTaskSchedulerFactory::createInRegistry();
+        assert(taskScheduler);
 
-        if (d_assembleMass.getValue())
+        if (d_parallelAssemblyIndependantMatrices.getValue() && taskScheduler && taskScheduler->getThreadCount() < 1)
```
Comment (fredroy) [inline Sofa/Component/LinearSystem/src/sofa/component/linearsystem/MatrixLinearSystem.inl]:
```suggestion
        if (d_parallelAssemblyIndependentMatrices.getValue() && taskScheduler && taskScheduler->getThreadCount() < 1)
```

File: Sofa/Component/LinearSystem/src/sofa/component/linearsystem/MatrixLinearSystem.inl [hunk 4]
```diff
@@ -116,26 +164,59 @@ void MatrixLinearSystem<TMatrix, TVector>::assembleSystem(const core::Mechanical
...
         {
-            contribute<Contribution::MASS>(mparams);
+            taskScheduler->init(0);
         }
 
-        if (d_assembleDamping.getValue())
-        {
-            contribute<Contribution::DAMPING>(mparams);
-        }
+        const simulation::ForEachExecutionPolicy execution = d_parallelAssemblyIndependantMatrices.getValue() ?
```
Comment (fredroy) [inline Sofa/Component/LinearSystem/src/sofa/component/linearsystem/MatrixLinearSystem.inl]:
```suggestion
        const simulation::ForEachExecutionPolicy execution = d_parallelAssemblyIndependentMatrices.getValue() ?
```

File: Sofa/Component/LinearSystem/src/sofa/component/linearsystem/MatrixLinearSystem.inl (hunk 2 above)
Comment (fredroy) [reply Sofa/Component/LinearSystem/src/sofa/component/linearsystem/MatrixLinearSystem.inl]: And should it be true by default ?

**Review by fredroy — CHANGES_REQUESTED** [review]

**Comment by alxbilger** [comment]
[ci-build][with-all-tests]

File: Sofa/Component/LinearSystem/src/sofa/component/linearsystem/MatrixLinearSystem.inl (hunk 2 above)
Comment (alxbilger) [reply Sofa/Component/LinearSystem/src/sofa/component/linearsystem/MatrixLinearSystem.inl]: to be set to false

**Review by fredroy — APPROVED** [review]

**Pull request merged**